Implement textDocument/documentSymbols

gabro opened the pull request:
Still WIP but I could use some feedback

The actual implementation of `DocumentSymbolsProvider` is just a port of @laughedelic's original work.

I've added:

- unit tests
- "integration" tests (see the slow suite)
- fallback logic for non-parsing buffers

gabro commented on metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala:
```diff
@@ -0,0 +1,157 @@
+  // FIXME(gabro): use a more sensible data structure
+  private var lastDocumentSymbolResult
+    : LinkedBlockingQueue[(AbsolutePath, List[DocumentSymbol])] =
+    new LinkedBlockingQueue(10)
```
I'm abusing a `LinkedBlockingQueue` to implement a LRU cache.

I didn't want to bring in additional dependencies, but I'm thinking to implement one myself.
Wdyt @olafurpg?

(the other obvious solution is to limit ourselves to 1 snapshot at the time, but it kinda sucks if you switch between two open non-parsing buffers)

olafurpg replied:
> the other obvious solution is to limit ourselves to 1 snapshot at the time, but it kinda sucks if you switch between two open non-parsing buffers

That's true, I was gonna suggest to limit it to one file. Instead we can keep `Map[Path, DocumentSymbol]` and remove entries on `didClose`

gabro replied:
I've just pushed a commit that uses `removeEldestEntry` from `LinkedHashMap` to implement a LRU cache.

olafurpg replied:
I think it's better to remove entries based on `textDocument/didClose` notifications so we don't have to pick a magic number.

gabro replied:
👍 I'll implement it that way (I wanted to limit the moving parts of `DocumentSymbolProvider`, but I'm uneasy with the magic number too)

laughedelic replied:
> That's true, I was gonna suggest to limit it to one file. Instead we can keep `Map[Path, DocumentSymbol]` and remove entries on `didClose`

👍 this was also the way it worked before, IIRC

gabro commented on metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala:
```diff
@@ -0,0 +1,157 @@
+            case (p, symbols) if p == path => symbols
+          }
+          .toList
+          .headOption
```
see comment below that explains this madness

gabro commented on tests/unit/src/test/resources/documentSymbol/example/AnonymousClasses.scala:
```diff
@@ -0,0 +1,3 @@
+/*example*/package example
+
+/*AnonymousClasses*/class AnonymousClasses {}
```
No ranges, yet.

@olafurpg The issue I have is that I don't have meaningful end positions in the `Semanticdbs.printTextDocument` method.

When I convert from `DocumentSymbol` to `SymbolOccurrence` I discard the end positions in order to make `printTextDocument` work.

Ideally, I would like to print the true end position of a document symbol. It's a meaningful test because ranges are used by editors to determine whether the cursor is "inside" a symbol and highlight the corresponding outline entry

olafurpg replied:
What if the symbol text inside the comment has the format `$KIND:$END_LINE`?

```scala
/*Namespace:20*/package example
 /*Class:10*/class AnonymousClasses {} 
```

gabro replied:
Same issue, the end line is equal to the start line in the `SymbolOccurrence` (or it would mess up with offset computation in `printTextDocument`)

olafurpg replied:
Can you elaborate? The end offset for `SymbolOccurrence.range` is not important, it can be the same as the start offset.

gabro replied:
Suppose I have a `DocumentSymbol` which spans until the end of the file (e.g. the symbol for a class definition).

When converting this symbol to a `SymbolOccurence`, I have to make the end position (line/char) identical to the end position, or the offset calculation will go wrong. Specifically, the issue is with these lines:

https://github.com/scalameta/metals/blob/d4199aa5f5a714e0b748e8e7e019e3049b97f21d/mtags/src/main/scala/scala/meta/internal/mtags/Semanticdbs.scala#L87-L94

If I pass in a `SymbolOccurrence` using the "correct" end position, the `.substring` call goes out of range.

olafurpg replied:
I left a comment for how to print the end line in `SymbolOccurrence.symbol`

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -361,4 +361,19 @@ object MetalsEnrichments extends DecorateAsJava with DecorateAsScala {
+          documentSymbol.getRange.getStart.getCharacter
+        )
+      ),
+      symbol = documentSymbol.getName,
```
Include `range.getEnd.getLine`

olafurpg replied:
this way we test the end line by printing the number in the output expect file, while discarding the end offset for the symbol occurrence range.

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -361,4 +361,19 @@ object MetalsEnrichments extends DecorateAsJava with DecorateAsScala {
   }
 
+  implicit class XtensionDocumentSymbol(documentSymbol: l.DocumentSymbol) {
+    def toSymbolOccurrence: s.SymbolOccurrence = s.SymbolOccurrence(
```
Thinking out loud. This seems only used in tests. I suspect the same applies for other methods in this file so it's not a big deal. Maybe we should have a similar utility file in tests?

gabro replied:
Done (only for `XtensionDocumentSymbol` right now). We can use to dump other test-only extensions.

olafurpg commented on tests/unit/src/main/scala/tests/TestingServer.scala:
```diff
@@ -176,6 +178,23 @@ final class TestingServer(
+    Debug.printEnclosing()
+    val abspath = toPath(filename)
+    val uri = abspath.toURI.toString
+    val extension = PathIO.extension(abspath.toNIO)
```
unused

olafurpg commented on tests/unit/src/main/scala/tests/TestingServer.scala:
```diff
@@ -176,6 +178,23 @@ final class TestingServer(
+    val abspath = toPath(filename)
+    val uri = abspath.toURI.toString
+    val extension = PathIO.extension(abspath.toNIO)
+    val text = abspath.readText
```
unused

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala:
```diff
@@ -0,0 +1,150 @@
+
+  private val snapshots = new TrieMap[AbsolutePath, List[DocumentSymbol]]()
+
+  private class SymbolTraverser(uri: String) {
```
`uri` is unused

gabro replied:
Right, it was used in a previous iteration, that I've reverted. Thanks for catching.

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala:
```diff
@@ -0,0 +1,150 @@
+  def discardSnapshot(path: AbsolutePath): Unit =
+    snapshots.remove(path)
+
+  private val snapshots = new TrieMap[AbsolutePath, List[DocumentSymbol]]()
```
nit, `TrieMap.empty` for consistency with rest of the codebase

olafurpg replied:
nit 2, I generally try to declare private mutable state  at the top of the class.

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -169,6 +170,7 @@ class MetalsLanguageServer(
       config.icons,
       statusBar
     )
+    documentSymbolProvider = new DocumentSymbolProvider(buffers)
```
Looks like this can be a `val` at the top of the class, `buffers` does not depend on `workspace`

gabro replied:
Yes, I've kept it like this for consistency, but I can use a `val`

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -0,0 +1,92 @@
+           |  "a": { }
+           |}
+           |/a/src/main/scala/a/Main.scala
+           |}package a
```
intended curly brace?

olafurpg replied:
see it's intended now, would be good to have something clearer

```scala
} // <- parse error
```

gabro replied:
Makes sense, I'll add it

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -0,0 +1,92 @@
+           |}
+           |/a/src/main/scala/a/Main.scala
+           |}package a
+           |import java.util.concurrent.Future // unused
```
Let's remove unnecessary code like imports, only keep relevant declarations.

olafurpg replied:
```scala
object Outer {
  class Inner
}
```

olafurpg replied:
the less code to copy-paste for expect tests the better

gabro replied:
👍 sounds good, I've based this one off `DefinitionSuite` but I agree it can be way shorter

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -0,0 +1,92 @@
+
+object DocumentSymbolSlowSuite extends BaseSlowSuite("documentSymbol") {
+
+  testAsync("documentSymbol") {
```
nit, I usually begin with a `"basic"` test suite when I have no idea for the name

olafurpg replied:
this test case stresses error handling however, so I would name it `"error"` instead

gabro commented:
Thank you so much for the review @olafurpg!

I've addressed all your comments and used your suggestion for printing the end line of symbols.

Ready for a second pass :)

olafurpg commented on tests/unit/src/main/scala/tests/MetalsTestEnrichments.scala:
```diff
@@ -0,0 +1,32 @@
+ *  Equivalent to scala.meta.internal.metals.MetalsEnrichments
+ *  but only for tests
+ */
+object MetalsTestEnrichments {
```
👍

olafurpg commented on tests/unit/src/main/scala/tests/MetalsTestEnrichments.scala:
```diff
@@ -0,0 +1,32 @@
+        )
+      ),
+      // include end line for testing purposes
+      symbol = s"${documentSymbol.getName}:${endRange.getLine + 1}",
```
s/getName/getKind/?

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -0,0 +1,66 @@
+      )
+      // fix the code to make it parse
+      _ <- server.didChange("a/src/main/scala/a/Main.scala") { text =>
+        text.replaceFirst("}", "")
```
Also works:
```scala
        """|object Outer {
           |  class Inner
           |}""".stripMargin
```

olafurpg replied:
whichever you think is clearer, I have no preference

gabro replied:
how about both?

Example:

```scala
/*example(Package):24*/package example

/*PatternMatching(Class):24*/class PatternMatching {
  /*some(Constant):4*/val some = Some(1)
  some match {
    case Some(number) =>
      number
  }

  // tuple deconstruction
  /*left(Constant):11*//*right(Constant):11*/val (left, right) = (1, 2)
  (left, right)

  // val deconstruction
  /*number1(Constant):16*/val Some(number1) =
    some
  println(number1)

  /*localDeconstruction(Method):23*/def localDeconstruction = {
    val Some(number2) =
      some
    number2
  }
}
```

olafurpg replied:
👍

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -0,0 +1,66 @@
+           |}""".stripMargin
+      )
+      // make the code unparseable again
+      _ <- server.didChange("a/src/main/scala/a/Main.scala")(text => text)
```
why is it unparseable again? I expected text to have the `}` removed

gabro replied:
I think it's a "bug" with `didChange`. It seems to always return the original buffer

olafurpg replied:
That makes sense, would be good to fix `didChange` so `text` has the buffer value

gabro replied:
I'll try :)

gabro replied:
Done 👍

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -0,0 +1,66 @@
+      )
+      // check that when closing the buffer, the snapshot is lost, and no symbols
+      // are found for unparseable code
+      _ <- server.didClose("a/src/main/scala/a/Main.scala")
```
👍

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSuite.scala:
```diff
@@ -0,0 +1,40 @@
+          val documentSymbols =
+            documentSymbolProvider.documentSymbols(file.file)
+
+          def toTextDocument(
```
Is this method needed?

gabro replied:
It doesn't have to be a method, I'll refactor it into a simple val. (it was more complicated before)

laughedelic commented on metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala:
```diff
@@ -0,0 +1,151 @@
+            symbolKind(currentNode),
+            currentNode.pos.toLSP,
+            currentNode.pos.toLSP
+          )
```
Is `currentNode.pos` the range of the whole definition (with the RHS)? I guess the 4th parameter is the `selectionRange`, which should correspond to the (more narrow) range of the identifier. This is not important, just saying that this can be improved later.

Also notice that there are more optional fields in this new `DocumentSymbol` type: `detail: String`, `deprecated: Boolean` and (the most interesting) `children: DocumentSymbol[]`. The old LSP's `SymbolInformation` declared that the range nesting and the `containerName` shouldn't be relied on for re-inferring the hierarchy, but editors used ranges anyway. Now there is this `children` field which makes this hierarchy explicit.

gabro replied:
Yes, I've taken some shortcuts in order to get this merged, but these are the things I would like to address next (in this order of priority):

- fallback to `SymbolInformation` for editors which don't support `DocumentSymbol` yet (e.g. vim-lsc)
- `children`
- `selectionRange`

I've agreed with @olafurpg to address those in future PR(s)

olafurpg replied:
I updated the PR so that

- `range` encloses the full range of the definition
- `selectionRange` encloses the definition name only.
- `detail` now includes the type ascription of the method (syntax written in source). It doesn't seem like vscode uses `detail` on the other hand.


We can add `deprecated` in a followup PR.

laughedelic replied:
Cool! 
I think that nested ranges would still work for `SymbolInformation` in those editors that don't support `DocumentSymbol` yet.

gabro replied:
Yeah, that's very much up to the editors.

For instance, the Code Outline extension for VSCode (which predates the native Outline View) used to use ranges for this. (VSCode's native view doesn't)

gabro commented:
Thank you for the review @laughedelic!

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala:
```diff
@@ -0,0 +1,215 @@
+      def newOwner(): Unit = {
+        continue(withNewOwner = true)
+      }
+      tree match {
```
I took the liberty to (hopefully) simplify the implementation. A bit of boilerplate can sometimes make it easier to and extend the code.

gabro commented on test-workspace/src/main/scala/example/User.scala:
```diff
@@ -5,6 +5,6 @@ import java.nio.file.Paths
 
 object User {
-  val sum: Int = 432
+  val sum: Int = ""
```
Accidental change?

olafurpg replied:
Yes, was experimenting. Forgot to undo the diff, but it doesn't matter much I suppose.

olafurpg replied:
Removed.

gabro commented on metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:
```diff
@@ -144,6 +144,24 @@ object MetalsEnrichments extends DecorateAsJava with DecorateAsScala {
+          s.getName,
+          s.getKind,
+          new l.Location(uri, s.getRange),
+          s.getDetail
```
I don't love the implicit detail <~> containerName correspondence, but since the LSP spec doesn't say much about it, I guess it's fine.

Let's keep in mind not to add fancier stuff to details, or this conversion will break.

olafurpg replied:
`containerName` now contains the enclosing fully qualified symbol which makes the output more readable in at least vim

<img width="724" alt="vim-document-symbol" src="https://user-images.githubusercontent.com/1408093/50220925-ed6e1080-0393-11e9-8db9-66015f3646e0.png">

gabro commented on metals/src/main/scala/scala/meta/internal/metals/MetalsServerConfig.scala:
```diff
@@ -88,6 +90,7 @@ object MetalsServerConfig {
           fileWatcher = FileWatcherConfig.auto,
           // window/logMessage output is always visible and non-invasive in vim-lsc
           statusBar = StatusBarConfig.logMessage,
+          documentSymbol = DocumentSymbolConfig.symbolInformation,
```
maybe add a comment here?

```scala
// vim-lsc does not yet support DocumentSymbol[] as a response for textDocument/documentSymbol
documentSymbol = DocumentSymbolConfig.symbolInformation,

```

olafurpg replied:
Done.

gabro commented:
@olafurpg fixed a failing test, then I think we can merge!

olafurpg commented on tests/unit/src/test/scala/tests/DocumentSymbolSlowSuite.scala:
```diff
@@ -39,7 +39,7 @@ object DocumentSymbolSlowSuite extends BaseSlowSuite("documentSymbol") {
         """|
            |/*Outer(Module):4*/object Outer {
-           |  /*Inner(Class):3*/class Inner
+           |  /*Outer.Inner(Class):3*/class Inner
```
Thanks! Forgot 😅

olafurpg commented:
The usual flaky tests on Windows, merging :shipit:
